File: dsa/courses/python-course/js-dsa-revision/0001-twoSum.js
/*
Difficulty: EASY 
No.1    https://leetcode.com/problems/two-sum/description/

Given an array of integers nums and an integer target, return indices of the two numbers such that they add up to target.
You may assume that each input would have exactly one solution, and you may not use the same element twice.
You can return the answer in any order.

Constraints:

2 <= nums.length <= 104
-109 <= nums[i] <= 109
-109 <= target <= 109
Only one valid answer exists.
*/

var twoSum = function (nums, target) {
  let left = 0;
  let right = left + 1;

  while (left < right) {
    let sum = nums[left] + nums[right];
    if (sum === target) {
      return [left, right];
    } else {
      right++;
      if (right === nums.length) {
        left++;
        right = left + 1;
      }
    }
  }
  return false;
};

console.log(twoSum([2, 7, 11, 15], 9)); // [0,1]
console.log(twoSum([3, 2, 4], 6)); // [1,2]
console.log(twoSum([3, 3], 6)); // [0,1]

File: leoncode/leetcode/0001-two-sum.py
# 1 - https://leetcode.com/problems/two-sum/

# Time: O(n)
# Space: O(n)
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        hm = {}

        for i, num in enumerate(nums):
            diff = target - num
            if diff in hm:
                return [hm[diff], i]
            hm[num] = i
        
        return []